File: DataModule/src/com/mobnetic/coinguardian/model/market/Bleutrade.java
package com.mobnetic.coinguardian.model.market;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;

import android.text.TextUtils;

import com.mobnetic.coinguardian.model.CheckerInfo;
import com.mobnetic.coinguardian.model.CurrencyPairInfo;
import com.mobnetic.coinguardian.model.Market;
import com.mobnetic.coinguardian.model.Ticker;

public class Bleutrade extends Market {
	
	private final static String NAME = "Bleutrade";
	private final static String TTS_NAME = NAME;
	private final static String URL = "https://bleutrade.com/api/v1/24h_trade_pair_info?market=%1$s_%2$s";
	private final static String URL_CURRENCY_PAIRS = "https://bleutrade.com/api/v1/trade_pairs";
	private final static HashMap<String, CharSequence[]> CURRENCY_PAIRS = new LinkedHashMap<String, CharSequence[]>();
	
	public Bleutrade() {
		super(NAME, TTS_NAME, CURRENCY_PAIRS);
	}
	
	@Override
	public String getUrl(int requestId, CheckerInfo checkerInfo) {
		return String.format(URL, checkerInfo.getCurrencyBase(), checkerInfo.getCurrencyCounter());
	}
	
	@Override
	protected void parseTicker(int requestId, String responseString, Ticker ticker, CheckerInfo checkerInfo) throws Exception {
		final String[] split = responseString.split(";");
		ticker.vol = parseDouble(split[5]);
		ticker.high = parseDouble(split[2]);
		ticker.low = parseDouble(split[3]);
		ticker.last = parseDouble(split[4]);
	}
	
	private double parseDouble(String doubleString) {
		return !TextUtils.isEmpty(doubleString) ? Double.parseDouble(doubleString) : 0;
	}
	
	// ====================
	// Get currency pairs
	// ====================
	@Override
	public String getCurrencyPairsUrl(int requestId) {
		return URL_CURRENCY_PAIRS;
	}
	
	@Override
	protected void parseCurrencyPairs(int requestId, String responseString, List<CurrencyPairInfo> pairs) throws Exception {
		final String[] currencyPairs = responseString.split("\n");
		for(int i=0; i<currencyPairs.length; ++i) {
			final String pairId = currencyPairs[i];
			final int underlineIndex = pairId.indexOf('_');
			pairs.add(new CurrencyPairInfo(
					pairId.substring(0, underlineIndex),
					pairId.substring(underlineIndex+1),
					pairId
				));
		}
	}
}

File: dataModule/src/main/java/com/aneonex/bitcoinchecker/datamodule/model/market/Bleutrade.kt
package com.aneonex.bitcoinchecker.datamodule.model.market

import com.aneonex.bitcoinchecker.datamodule.model.CheckerInfo
import com.aneonex.bitcoinchecker.datamodule.model.CurrencyPairInfo
import com.aneonex.bitcoinchecker.datamodule.model.Market
import com.aneonex.bitcoinchecker.datamodule.model.Ticker
import org.json.JSONArray
import org.json.JSONObject

class Bleutrade : Market(NAME, TTS_NAME, null) {
    override fun getUrl(requestId: Int, checkerInfo: CheckerInfo): String {
        return String.format(URL, checkerInfo.currencyPairId)
    }

    @Throws(Exception::class)
    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker, checkerInfo: CheckerInfo) {
        val resultObject = jsonObject["result"]
        val resultsJsonObject = if (resultObject is JSONArray) {
            resultObject.getJSONObject(0)
        } else {
            resultObject as JSONObject
        }
        ticker.bid = resultsJsonObject!!.getDouble("Bid")
        ticker.ask = resultsJsonObject.getDouble("Ask")
        ticker.last = resultsJsonObject.getDouble("Last")
    }

    @Throws(Exception::class)
    override fun parseErrorFromJsonObject(requestId: Int, jsonObject: JSONObject, checkerInfo: CheckerInfo?): String? {
        return jsonObject.getString("message")
    }

    // ====================
    // Get currency pairs
    // ====================
    override fun getCurrencyPairsUrl(requestId: Int): String? {
        return URL_CURRENCY_PAIRS
    }

    @Throws(Exception::class)
    override fun parseCurrencyPairsFromJsonObject(requestId: Int, jsonObject: JSONObject, pairs: MutableList<CurrencyPairInfo>) {
        val resultsJsonArray = jsonObject.getJSONArray("result")
        for (i in 0 until resultsJsonArray.length()) {
            val pairJsonObject = resultsJsonArray.getJSONObject(i)
            val pairId = pairJsonObject.getString("MarketName")
            val currencyBase = pairJsonObject.getString("MarketCurrency")
            val currencyCounter = pairJsonObject.getString("BaseCurrency")
            pairs.add(CurrencyPairInfo(
                    currencyBase,
                    currencyCounter,
                    pairId
            ))
        }
    }

    companion object {
        private const val NAME = "Bleutrade"
        private const val TTS_NAME = NAME
        private const val URL = "https://bleutrade.com/api/v2/public/getticker?market=%1\$s"
        private const val URL_CURRENCY_PAIRS = "https://bleutrade.com/api/v2/public/getmarkets"
    }
}